Applications/Spire/Source/OrderImbalanceIndicator/FilteredOrderImbalanceIndicatorModel.cpp

```diff
@@ -0,0 +1,114 @@
+#include "Spire/OrderImbalanceIndicator/FilteredOrderImbalanceIndicatorModel.hpp"
+
+using Filter = Spire::FilteredOrderImbalanceIndicatorModel::Filter;
+using namespace Nexus;
+using namespace Spire;
+
+FilteredOrderImbalanceIndicatorModel::FilteredOrderImbalanceIndicatorModel(
+  std::shared_ptr<OrderImbalanceIndicatorModel> source_model,
+  std::vector<Filter> filters)
+  : m_source_model(std::move(source_model)),
+    m_filters(std::move(filters)) {}
+
+QtPromise<std::vector<Nexus::OrderImbalance>>
+    FilteredOrderImbalanceIndicatorModel::load(
+    const TimeInterval& interval) {
+  return m_source_model->load(interval).then([=] (const auto& imbalances) {
+    return filter_imbalances(imbalances.Get());
+  });
+}
+
+SubscriptionResult<boost::optional<Nexus::OrderImbalance>>
+    FilteredOrderImbalanceIndicatorModel::subscribe(
+    const OrderImbalanceSignal::slot_type& slot) {
+  return m_source_model->subscribe([=] (const auto& imbalance) {
```

Comment:
Same issue here, risk of a dangling reference.

Reply:
In the first example, the call to load() creates a promise, and the call to then() creates another. The issue being the original load() promise will go out of scope if it's not stored.

But how does the dangling reference issue apply here if the sole promise is being returned?

Reply:
Add a test case as follows:

* Construct a local model L (as a shared_ptr).
* Construct a filtered model F (as a unique_ptr) and pass as an argument L.
* Call F->subscribe(P) for some arbitrary value P.
* Call F.reset(), destroying the filtered model.
* Call L.publish(M).
* Expected result, no crash.

Reply:
So is the expectation that it won't crash in its current form? Or that the correct implementation won't crash?

It's passing in its current form but is it possible that reset isn't actually deleting anything so the filtered model still exists in memory?

Reply:
Push the test.

Reply:
It's the last test in the filtered imbalance model tester file. I kept the placeholder names to make reference the PR notes easier.

I temporary copied the test model into the tester source because it's in the other cache branch, but if I need it I'll move it into it's own files.

Reply:
Are you running these tests in debug mode or release mode?

Reply:
Debug.

Reply:
I can't test this right now but if you go over the test it should be clear that the application should fail.

Reply:
I think my original implementation may be partially correct. Following the call stack for the source (test) model's load method, it calls then() (link below), which constructs another promise from the original promise; so storing a handle to the original promise may not be required.

[QtPromise Link](https://github.com/eidolonsystems/nexus/blob/d3c5a8374c344c9a598cc575d7875b0000d40405/Applications/Spire/Include/Spire/Spire/QtPromise.hpp#L174)

This doesn't fix the issue with the filter model being destroyed, though.

Reply:
I had to add a wait() call for the filtered model's load() promise. I suspect the test was ending, then the test model would be destroyed, so the test model's load method didn't return a value to the filtered model for the crash to happen.